creator-node/src/fileManager.js

```diff
@@ -102,50 +104,98 @@ async function saveFileToIPFSFromFS (req, srcPath, fileType, sourceFile, transac
   }
 
   // If file not already available on local INode, fetch from IPFS.
   if (fileBuffer === null) {
-    req.logger.info(`Attempting to get ${multihash} from IPFS`)
+    req.logger.debug(`Attempting to get ${multihash} from IPFS`)
     let output
     try {
-      output = await ipfs.get(multihash)
+      output = await Utils.ipfsGet(multihash, req, 1000)
+      if (output.length !== 1) throw new Error(`provided multihash ${multihash} must map to 1 file`)
+      fileBuffer = output[0].content
+      req.logger.debug(`retrieved file for multihash ${multihash} from path ${output[0].path}`)
+      // Write file to disk.
+      await writeFile(expectedStoragePath, fileBuffer)
+      req.logger.info(`wrote file to ${expectedStoragePath}, obtained via ipfs get`)
     } catch (e) {
-      throw new Error(`Failed to retrieve file for multihash ${multihash} from IPFS`)
+      req.logger.info(`Failed to retrieve file for multihash ${multihash} from IPFS`)
     }
-    if (output.length !== 1) throw new Error('Audius track segment multihash must map to 1 file')
-    fileBuffer = output[0].content
-    req.logger.info(`retrieved file for multihash ${multihash} from path ${output[0].path}`)
   }
 
-  // Write file to disk.
-  const storagePath = path.join(req.app.get('storagePath'), multihash)
-  assert.strictEqual(storagePath, expectedStoragePath)
-  req.logger.info(`writing file to ${storagePath}...`)
-  await writeFile(storagePath, fileBuffer)
-  req.logger.info(`wrote file to ${storagePath}`)
+  // if file is still null, try to fetch from other cnode gateways with onlyFS=true
+  if (fileBuffer === null) {
+    try {
+      let response
+      // ..replace(/\/$/, "") removes trailing slashes
+      req.logger.debug(`Attempting to fetch multihash ${multihash} by racing replica set endpoints`)
+      const urls = gatewaysToTry.map(endpoint => `${endpoint.replace(/\/$/, '')}/ipfs/${multihash}`)
+
+      // TODO make this more parallel
+      for (let index = 0; index < urls.length; index++) {
+        const url = urls[index]
+        try {
+          const resp = await axios({
+            method: 'get',
+            url,
+            responseType: 'stream',
+            timeout: 5000, /* ms */
+            params: { onlyFS: true }
+          })
+          if (resp.data) {
+            response = resp
+            break
+          }
+        } catch (e) {
+          continue
+        }
+      }
+
+      if (!response || !response || !response.data) {
+        throw new Error(`Couldn't find files on other creator nodes via promiseRace`)
+      }
 
-  req.logger.info(`\nAdded file: ${multihash} at ${storagePath}`)
-  return storagePath
+      await response.data.pipe(fs.createWriteStream(expectedStoragePath))
+      req.logger.info(`wrote file to ${expectedStoragePath}`)
+    } catch (e) {
+      throw new Error(`Failed to retrieve file for multihash ${multihash} from other creator node gateways: ${e.message}`)
```

Comment:
just to double check, if an error is thrown here, is it properly being handled in the node sync?

Reply:
yes there's a try-catch in nodesync